src/main/java/org/jitsi/jicofo/ChannelAllocator.java

```diff
@@ -0,0 +1,688 @@
+/*
+ * Jicofo, the Jitsi Conference Focus.
+ *
+ * Copyright @ 2015 Atlassian Pty Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jicofo;
+
+import net.java.sip.communicator.impl.protocol.jabber.extensions.colibri.*;
+import net.java.sip.communicator.impl.protocol.jabber.extensions.jingle.*;
+import net.java.sip.communicator.impl.protocol.jabber.extensions.jitsimeet.*;
+import net.java.sip.communicator.impl.protocol.jabber.jinglesdp.*;
+import net.java.sip.communicator.service.protocol.*;
+
+import org.jitsi.impl.protocol.xmpp.extensions.*;
+import org.jitsi.jicofo.discovery.*;
+import org.jitsi.jicofo.util.*;
+import org.jitsi.protocol.xmpp.*;
+import org.jitsi.protocol.xmpp.colibri.*;
+import org.jitsi.service.neomedia.*;
+import org.jitsi.util.*;
+
+import java.util.*;
+
+/**
+ * The class is a thread that does the job of allocating Colibri channels on
+ * the bridge and invites participant with Jingle 'session-initiate'. Also
+ * bridge election happens here with some help of {@link BridgeSelector}.
+ *
+ * @author Pawel Domas
+ */
+public class ChannelAllocator implements Runnable
+{
+    /**
+     * Error code used in {@link OperationFailedException} when there are no
+     * working videobridge bridges.
+     * FIXME: consider moving to OperationFailedException ?
+     */
+    private final static int BRIDGE_FAILURE_ERR_CODE = 20;
+
+    /**
+     * The logger instance used in this class.
+     */
+    private final static Logger logger
+        = Logger.getLogger(ChannelAllocator.class);
+
+    /**
+     * Parent {@link JitsiMeetConference}.
+     */
+    private final JitsiMeetConference meetConference;
+
+    /**
+     * <tt>ColibriConference</tt> instance used by this thread to allocate
+     * channels on the bridge.
+     */
+    private final ColibriConference colibriConference;
+
+    /**
+     * A participant that is to be invited by this instance to the conference.
+     */
+    private final Participant newParticipant;
+
+    /**
+     * First argument stands for "start audio muted" and the second one for
+     * "start video muted". The information is included as a custom extension in
+     * 'session-initiate' sent to the user.
+     */
+    private final boolean[] startMuted;
```

Comment:
I'm starting to feel uncomfortable with the startMuted array in its multiple incarnations. I've seen it in at least JitsiMeetConference and ChannelAllocator. And there were multiple methods calls with it as an arguments.

1. It's cryptic because you have to know what's what at which index. (I've used the approach as well so I'm not stating I'm a saint.)

2. JitsiMeetConference.hasToStartMuted checks for null but  AbstractOperationSetJingle.createInviteIQ doesn't. In practice, you're unlikely to get to a null with our current usage. But it's still not pretty.

I'd suggest an enum, maybe flag-style or something.